### PR TITLE
Upgrade to Hive 3.0.0

### PR DESCRIPTION
**Pull request by imjalpreet**
This PR includes the base changes required for upgrading to Hive 3. This will help in supporting and bringing a number of interesting features in Hive 3 to Presto in the future.

depends on https://github.com/facebookexternal/presto-facebook/pull/1450

### CODE REVIEW

**Comment by aweisberg** [comment]
Added you as a contributor. Not sure if that will let you restart tests yourself. I will need to force push to update just to get the FB integration going.

File: pom.xml
Comment (pettyjamesm) [inline pom.xml:584]: Should we just jump directly to 3.0.0-3 in this PR or re-apply the changes from https://github.com/prestodb/presto/pull/15796 after this one is merged?

Comment (aweisberg) [reply pom.xml:584]: Yeah I'll update it

File: pom.xml
Comment (pettyjamesm) [inline pom.xml:584]: Missing `/`, should be `</version>`

Comment (aweisberg) [reply pom.xml]: Thanks again. Fixed.

**Comment by aweisberg** [comment]
@imjalpreet this is the same as the last PR just again modulo the one change I made to the Hive version (3.0.0-3 instead of 3.0.0)? We don't need to do a full review?

**Comment by aweisberg** [comment]
The collection delim fixes look correct to me. I added the "Cherry-pick of" to it. It would be nice to also have those for the "Update to HIve 3.0.0". Generally linking context a bit so people don't have to search is low effort and helpful.

I am going to test this now against the queries we had issues with and we can see if the issue is resolved.

**Comment by imjalpreet** [comment]
@aweisberg Yes this is the same PR as the earlier one. It just includes the fix for collection delimiter for Hive versions < 3.0.0.

**Comment by aweisberg** [comment]
I was able to test it on one of the failed queries and it works correctly now. The tables are gone for the other two, but I am confident that we fixed the underlying issue for each query.

**Comment by aweisberg** [comment]
Everything passed, but failed on a flakey test.
I realized if you have multiple compute engines this won't work if it writes the correct version and the other engines don't support both.

**Comment by aweisberg** [comment]
Trino doesn't appear to do anything special. I can't find `colelction.delim` in additional places in `trino` or `trino-hive-apache`.

It seems like a real concern to me, but maybe we just release note it? I imagine not many people are using TEXTFILE as an output format in Presto.

**Comment by pettyjamesm** [comment]
> Trino doesn't appear to do anything special. I can't find `colelction.delim` in additional places in `trino` or `trino-hive-apache`.
> 
> It seems like a real concern to me, but maybe we just release note it? I imagine not many people are using TEXTFILE as an output format in Presto.

I wonder how much work it would be to add a config option to emit both names or maybe just the "legacy" name if enabled? I suspect there is a relatively small number of users using text formats and also using other compute engines, but I'm sure it exists and they'll be upset if they have no recourse.

**Comment by aweisberg** [comment]
> I wonder how much work it would be to add a config option to emit both names or maybe just the "legacy" name if enabled? I suspect there is a relatively small number of users using text formats and also using other compute engines, but I'm sure it exists and they'll be upset if they have no recourse.

I was thinking about adding both, but is it the case that it will error out on unrecognized options if both are present?

**Review by aweisberg — APPROVED** [review]

**Comment by pettyjamesm** [comment]
> I was thinking about adding both, but is it the case that it will error out on unrecognized options if both are present?

Can't say that I know... seems like that would depend on the SerDe or other query engine in question, which makes it kind of a black box. It _does_ seem kind of wonky to set both properties by default and keep propagating that silly typo into the future, so making the creation configurable but accepting either on when reading seems like the best compromise to me.

**Comment by aweisberg** [comment]
I looked at the hive source and it seems like it does track what the list of supported properties are in an annotation, but not to error.

There is one case where it compares props from two different tables and this would break in an environment that mixes them.

TBH not sure what to do here, but the urge to ship it is strong.

**Comment by pettyjamesm** [comment]
> TBH not sure what to do here, but the urge to ship it is strong.

I agree, it's hard to know what the right thing to do here is in advance. It's probably the right call to just move forward, document the potential incompatibility, and see who comes forward with issues afterwards. Once we get a sense for who this broke we'll have a better answer for how to mitigate it- and it does seem like something that can be mitigated.

**Comment by rschlussel** [comment]
> > TBH not sure what to do here, but the urge to ship it is strong.
> 
> I agree, it's hard to know what the right thing to do here is in advance. It's probably the right call to just move forward, document the potential incompatibility, and see who comes forward with issues afterwards. Once we get a sense for who this broke we'll have a better answer for how to mitigate it- and it does seem like something that can be mitigated.

I agree.  If we know the use cases it broke for, we'll be able to know what solution is best. For now at least not breaking reads seems sufficient.

**Review by rschlussel — APPROVED** [review]

**Review by pettyjamesm — APPROVED** [review]

**Comment by rschlussel** [comment]
@aweisberg or @imjalpreet  can you add the release note information to the PR description.

**Comment by imjalpreet** [comment]
@rschlussel @aweisberg I have updated the description with the release note, please let me know in case anything else needs to be added.

**Comment by shangxinli** [comment]
@imjalpreet, Can you explain why the dependency on parquet-xxx is removed in this change? By doing that, we are only using the Parquet via Hive but not directly using Parquet anymore. I am working on https://github.com/prestodb/presto/pull/14960, which is to upgrade Parquet version to 1.11.x, but surprisingly that doesn't work anymore because of this change.  FYI @vkorukanti @zhenxiao @highker

**Comment by imjalpreet** [comment]
> @imjalpreet, Can you explain why the dependency on parquet-xxx is removed in this change? By doing that, we are only using the Parquet via Hive but not directly using Parquet anymore. I am working on #14960, which is to upgrade Parquet version to 1.11.x, but surprisingly that doesn't work anymore because of this change. FYI @vkorukanti @zhenxiao @highker

@shangxinli The parquet dependencies were only added here until the hive was upgraded. This could be done since the earlier version of the shaded version of Hive (we were using 1.2.2-2) had Parquet 1.6.0 and since Apache had changed the namespace of Parquet classes in the recent versions it allowed having both new and old parquet classes in the classpath. 

With the upgrade to Hive 3, we don't need to separately include the parquet dependencies as we get them along with Hive itself. 

Since we maintain shaded version of Hive for Presto, you can upgrade parquet to 1.11.x in [presto-hive-apache](https://github.com/prestodb/presto-hive-apache/blob/master/pom.xml#L56) and we can release that change and start using that. I think there is already a [PR](https://github.com/prestodb/presto-hive-apache/pull/46) in progress to upgrade to 1.11.0.

Let me know in case you have any questions.

**Comment by shangxinli** [comment]
Thanks @imjalpreet for the quick response! I can understand the reason and it is a good effort to bring into one version. 

What is the release frequency of presto-hive-apache? It seems upgrading Parquet in Presto becomes more difficult as we have to 1) change presto-hive-apache 2) Wati till the release presto-hive-apache 3) change Presto.  I already upgraded Iceberg to 1.11.1 and I also opened a ticket to upgrade to Parquet 1.12.0. When Iceberg connector catches up upgrading, that means we have to go through #1, #2, #3 very often. 

Also, with this change when upgrading presto-hive-apache in Presto, that could *implicitly* upgrade Parquet to newer version. For PR https://github.com/prestodb/presto/pull/14960, several committers already have concerns and want us to wait till deployment in production. I am not sure if we still have this path of control if we *implicitly* upgrade Parquet in the future. 

Are you open to add back the direct dependency on Parquet? If yes, I can add them back in my PR https://github.com/prestodb/presto/pull/14960.

**Comment by shangxinli** [comment]
Any thoughts @imjalpreet?

**Comment by imjalpreet** [comment]
@shangxinli Sorry I missed your comment. Please find my response below.

> What is the release frequency of presto-hive-apache? It seems upgrading Parquet in Presto becomes more difficult as we have to 1) change presto-hive-apache 2) Wait till the release presto-hive-apache 3) change Presto.  I already upgraded Iceberg to 1.11.1 and I also opened a ticket to upgrade to Parquet 1.12.0. When Iceberg connector catches up upgrading, that means we have to go through steps 1), 2), 3) very often. 
> 
There isn't a set release frequency for presto-hive-apache, as far as I am aware it can be released whenever there is a change required or update required. @rschlussel can confirm.

> Also, with this change when upgrading presto-hive-apache in Presto, that could *implicitly* upgrade Parquet to newer version. For PR https://github.com/prestodb/presto/pull/14960, several committers already have concerns and want us to wait till deployment in production. I am not sure if we still have this path of control if we *implicitly* upgrade Parquet in the future. 
> 
I assume those releases can be coordinated. Although @rschlussel @aweisberg would be better to comment on that.
> 
> 
> Are you open to add back the direct dependency on Parquet? If yes, I can add them back in my PR https://github.com/prestodb/presto/pull/14960. 
> 

I don't think it is possible now to add the direct dependency to Parquet. It was possible earlier because previous version of Hive used to have a different classpath for Parquet dependencies, so it was possible to have different versions of Parquet present in the classpath.
But Hive > 3 has the new classpath for the parquet dependencies which Apache had modified in the recent versions of Parquet. 

The point I am trying to make is even if you add a direct dependency to Parquet and exclude the Parquet dependencies coming from presto-hive-apache, on any upgrade of Parquet in Presto you will have to make some changes in presto-hive-apache since there can be breaking changes with an upgrade and hive read and write path would expect the older versions of Parquet classes since it had been build with that version.

Unfortunately there is no way to skip this. 

Let me know in case you have any concerns.

**Comment by shangxinli** [comment]
Thanks @imjalpreet! If that is the only option, let's go with that. The only thing is we have a dependency on "upgrading to 1.11.x" to open source Parquet Column Index(PR https://github.com/prestodb/presto/issues/15454) and Column Encryption  PRhttps://github.com/prestodb/presto/issues/12048). It would be great if you can help to expedite the [PR](https://github.com/prestodb/presto-hive-apache/pull/46).

**Comment by imjalpreet** [comment]
@aweisberg @rschlussel would you be able to help on this?

**Comment by imjalpreet** [comment]
> @imjalpreet, I see [PR46](https://github.com/prestodb/presto-hive-apache/pull/46) has been merged. For the next steps, I guess we need to produce a new version of 'presto-hive-apache' and make a change in Presto. Is it something you will take care of, or you want me to do something? Let me know. 

@shangxinli First requirement will be to release the new version of presto-hive-apache and once that is done we can work on the changes to update the version here and if any other changes that arise due to that.

For releasing the new version of presto-hive-apache, @aweisberg @rschlussel can you please help?

**Comment by rschlussel** [comment]
yeah, i can do the release

**Comment by rschlussel** [comment]
presto-hive-apache 3.0.0-4 should now be released

**Comment by imjalpreet** [comment]
@rschlussel thanks for the help.

**Pull request merged**
